src/multiexp.rs

```diff
@@ -0,0 +1,278 @@
+use crate::curve_jac_ext::CurveJacExt;
+use group::{ff::PrimeField, Group as _};
+use rayon::{
+    prelude::{IndexedParallelIterator, ParallelIterator},
+    slice::ParallelSlice,
+};
+
+/// Performs a multi-exponentiation operation.
+///
+/// This function will panic if coeffs and bases have a different length.
+///
+/// This will use multithreading if beneficial.
+pub fn best_multiexp<C: CurveJacExt>(
+    coeffs: &[C::Scalar],
+    bases: &[C],
+) -> C::ExtendedJacobianCoordinates {
+    let n = coeffs.len();
+    assert_eq!(n, bases.len());
+
+    let num_threads = num_cpus::get();
+    if n > num_threads && n > 32 {
+        let chunk = n / num_threads;
+        let results: Vec<C::ExtendedJacobianCoordinates> = coeffs
+            .par_chunks(chunk)
+            .zip(bases.par_chunks(chunk))
+            .map(|(coeffs, bases)| multiexp_serial(coeffs, bases))
+            .collect();
+        results.iter().fold(C::jac_ext_identity(), |a, b| a + *b)
+    } else {
+        small_multiexp(coeffs, bases)
+    }
+}
+
+pub(crate) fn multiexp_serial<C: CurveJacExt>(coeffs: &[C::Scalar], bases: &[C]) -> C::ExtendedJacobianCoordinates {
+    let c = if bases.len() < 4 {
+        1
+    } else if bases.len() < 32 {
+        3
+    } else {
+        (bases.len() as f64).ln().ceil() as usize
+    };
+    let mut buckets: Vec<Vec<Bucket<C>>> = vec![vec![Bucket::None; (1 << c) - 1]; (256 / c) + 1];
+
+    buckets
+        .iter_mut()
+        .enumerate()
+        .rev()
+        .map(|(i, bucket)| {
+            for (coeff, base) in coeffs.iter().zip(bases.iter()) {
```

Comment:
The map and for are inefficient, you loop twice over the data.

You should loop over the coeff+base and directly mutate/add_assign in the matching bucket